src/libtrx/game/items/walkable.c

```diff
@@ -0,0 +1,167 @@
+#include "game/items/walkable.h"
+
+#include "game/game_buf.h"
+#include "game/items.h"
+#include "game/objects/general/drawbridge.h"
+#include "game/objects/general/trapdoor.h"
+#include "game/objects/vars.h"
+#include "game/rooms.h"
+#include "log.h"
+#include "memory.h"
+#include "vector.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+static SECTOR *M_GetItemPitSector(const ITEM *item, XYZ_32 pos);
+static void M_InsertSorted(WALKABLE **walkables, WALKABLE *node);
+void M_Add(int16_t item_num, const ITEM *item, XYZ_32 pos);
+
+static SECTOR *M_GetItemPitSector(const ITEM *const item, const XYZ_32 pos)
+{
+    int16_t room_num = item->room_num;
+    SECTOR *sector = Room_GetSector(pos.x, pos.y, pos.z, &room_num);
+    // LOG_DEBUG("room_num: %d", room_num);
+    return Room_GetPitSector(sector, pos.x, pos.z);
+}
+
+static void M_InsertSorted(WALKABLE **walkables, WALKABLE *node)
+{
+    while (*walkables && (*walkables)->pos.y >= node->pos.y) {
+        walkables = &(*walkables)->next;
+    }
+
+    node->next = *walkables;
+    *walkables = node;
+}
+
+void M_Add(const int16_t item_num, const ITEM *const item, const XYZ_32 pos)
+{
+    SECTOR *sector = M_GetItemPitSector(item, pos);
+
+    // Check if the walkable is already in the sector.
+    WALKABLE **walkables = &sector->walkable;
+    while (*walkables) {
+        if ((*walkables)->item_num == item_num) {
+            LOG_DEBUG("item_num: %d already exists in this sector", item_num);
+            return;
+        }
+        walkables = &(*walkables)->next;
+    }
+
+    LOG_DEBUG(
+        "Add %d; xyz: %d %d %d; floor.height: %d; ceiling.height: %d", item_num,
+        pos.x, pos.y, pos.z, sector->floor.height, sector->ceiling.height);
+    WALKABLE *node = GameBuf_Alloc(sizeof(WALKABLE), GBUF_WALKABLES);
+    node->item_num = item_num;
+    node->pos = pos;
+    node->next = NULL;
+    M_InsertSorted(&sector->walkable, node);
+}
+
+void Walkable_Add(const int16_t item_num)
+{
+    ITEM *item = Item_Get(item_num);
+    if (item->object_id == O_DRAWBRIDGE) {
+        LOG_DEBUG(
+            "item %d OG pos: %d %d %d", item_num, item->pos.x, item->pos.y,
+            item->pos.z);
+        VECTOR *drawbridge_pos = Vector_Create(sizeof(XYZ_32));
+        Drawbridge_GetSectorPositions(item, drawbridge_pos);
+        for (int32_t i = 0; i < drawbridge_pos->count; i++) {
+            M_Add(
+                item_num, item, *(const XYZ_32 *)Vector_Get(drawbridge_pos, i));
+        }
+        Vector_Free(drawbridge_pos);
+        drawbridge_pos = nullptr;
+    } else if (
+        item->object_id == O_TRAPDOOR_TYPE_1
+        || item->object_id == O_TRAPDOOR_TYPE_2
+        || item->object_id == O_TRAPDOOR_TYPE_3) {
+        LOG_DEBUG(
+            "Trapdoor item %d OG pos: %d %d %d", item_num, item->pos.x,
+            item->pos.y, item->pos.z);
+        VECTOR *trapdoor_pos = Vector_Create(sizeof(XYZ_32));
+        Trapdoor_GetSectorPositions(item, trapdoor_pos);
```

Comment:
WDYT about making this function a part of the `OBJECT` structure, as a function pointer, similar to for example `draw` or `control`? This way we don't have to hardcode object IDs here.